chain/events/filter/resolver.go

```diff
@@ -0,0 +1,118 @@
+package filter
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/golang-lru/v2/expirable"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/abi"
+
+	"github.com/filecoin-project/lotus/chain/types"
+)
+
+type (
+	// AddressResolver resolves an actor ID to its address at a given TipSet, or the latest TipSet if nil.
+	AddressResolver func(context.Context, abi.ActorID, *types.TipSet) (address.Address, error)
```

Comment:
So, this needs to be specified a bit more. Unfortunately, in the post-FEVM land, an actor can have _both_ a "delegated" (f4) address AND an "actor" (f2) address. So basically, when "resolving" an ID address, the possibilities are:
ID -> f1 (secp address)
ID -> f3 (BLS address)
ID -> f2 ("actor" address, this is true for miners, multisigs, etc.)
ID -> f2 AND f4 ("actor" and "delegated" address, I think this is true of EVM actors)
(and maybe) ID -> f4 (just a "delegated" address, this is true of...EthAccounts? I'm not sure of this one.)

So, when doing any resolution starting from an ID address, you need to decide which one you want to prefer if _both_ are present. My instinct is to say that you should prefer the f4 address in this case (it exists for Ethereum-like behaviour, so we might as well use it for Ethereum events, while falling back on robust f2 addresses for everything else). But this is probably more a @Stebalien question.

(This wasn't a problem before, because ONLY Ethereum-like things were emitting events, so the AddressResolver could exclusively rely on f4 addresses).